README.md
# @senntyou/utils

工具函数、快捷对象、工具集等

```
npm install @senntyou/utils --save
```

```
import { isMobile, isWeiXin, ... } from '@senntyou/utils';
```

## isMobile：是否是移动端

`type: bool`

## isWeiXin：是否是微信浏览器

`type: bool`

## isIos：是否是 ios 终端

`type: bool`

## isMiniProgram：是否是微信小程序 web-view

`type: bool`

## getDate：获取一个时间对象

`type: function`

```
const date = getDate(d);

const {
  d,          // Date 对象实例
  year,       // 年
  month,      // 月
  day,        // 日
  week,       // 周 0-6
  hours,      // 小时
  minutes,    // 分钟
  seconds,    // 秒
  date,       // 日期 2019-01-01
  time,       // 时间 01:01:01
  dateTime,   // 日期时间 2019-01-01 01:01:01
} = date;
```

- `@param/d`: `type: Date` `default: 当前` 获取某个时刻的时间数据
- `@return`: `type: {}` 时间数据

## now：当前时间数据

```
const {
  d,          // Date 对象实例
  year,       // 年
  month,      // 月
  day,        // 日
  week,       // 周 0-6
  hours,      // 小时
  minutes,    // 分钟
  seconds,    // 秒
  date,       // 日期 2019-01-01
  time,       // 时间 01:01:01
  dateTime,   // 日期时间 2019-01-01 01:01:01
} = now;
```

## refreshNow: 刷新当前时间数据

`type: function`

```
refreshNow(d)
```

- `@param/d`: `type: Date` `default: 当前` 刷新到哪一个时间，默认是当前

## urlParams：url 参数

`type: {}`

```
'url?key1=1&key2=2&key3='    =>
{
  key1: '1',
  key2: '2',
  key3: '',
}
```

## fillZero：填充 0

`type: function`

```
fillZero(num, length)
```

- `@param/num`: `type: number` `default: 0` 待处理的数字
- `@param/length`: `type: number` `default: 2` 填充长度
- `@return`: `type: string` 填充后的字符串

```
fillZero(1)        // '01'
fillZero(1, 4)        // '0001'
fillZero(12, 4)        // '0012'
```

## reloadTitle：重新加载微信浏览器的标题，hack 在微信 webview 中无法修改 document.title 的情况

`type: function`

```
reloadTitle(title)
```

- `@param/title`: `type: string` 标题

## makeUrlSearch：make url params search

`type: function`

```
makeUrlSearch(params)
```

- `@param/params`: `type: {}` 参数
- `@return`: `type: string` url search

```
makeUrlSearch({key1: 1, key2: 2})                // 'key1=1&key2=2'
```

## parseUrlSearch：parse url search

`type: function`

```
const params = parseUrlSearch(search)
```

- `@param/search`: `type: string` 参数
- `@return`: `type: {}` params

```
parseUrlSearch('?key1=1&key2=2')                 // {key1: 1, key2: 2}
```

## posToDoc：元素距离文档顶部的位置信息

`type: function`

```
posToDoc(el)
```

- `@param/params`: `type: dom` 元素
- `@return`: `type: {}` 位置信息

```
posToDoc(el)                 // { x, y }
```

## loadImages：加载图片快捷函数

`type: function`

```
loadImages(images, cb)
```

- `@param/images`: `type: string/string[]` 图片集合
- `@param/cb`: `type: function` 回调函数

```
loadImages([url1, url2, ...], () => {
  // 所有图片都加载完成
})
```

## numOfDate：把日期转成数字

`type: function`

```
numOfDate(date)
```

- `@param/date`: `type: string` 日期
- `@return`: `type: number`

```
numOfDate('2019-01-01')                // 20190101
```

## numOfTime：把时间转成数字

`type: function`

```
numOfTime(time)
```

- `@param/time`: `type: string` 时间
- `@return`: `type: number`

```
numOfTime('12:12:12')                  // 121212
```

## numOfDateTime：把日期时间转成数字

`type: function`

```
numOfDateTime(dateTime)
```

- `@param/dateTime`: `type: string` 日期时间
- `@return`: `type: number`

```
numOfDateTime('2019-01-01 12:12:12')   // 20190101121212
```

## distanceOfDates：返回两个日期之间的距离，以天计

`type: function`

```
distanceOfDates(start, end)
```

- `@param/start`: `type: string` 开始日期
- `@param/end`: `type: string` 结束日期
- `@return`: `type: number` 距离

```
distanceOfDates('2019-01-01', '2019-01-10')      // 9
distanceOfDates('2019-01-10', '2019-01-02')      // -8
```

## readableTime：格式化为可读时间（n 秒前、n 分钟前、n 小时前、n 天前、n 个月前、n 年前）

`type: function`

```
readableTime(time)
```

- `@param/time`: `type: string` 日期
- `@return`: `type: string` 可读时间

```
readableTime('2019-01-01 01:01:01')              // 12 秒前
```

## dateByInterval：获取距离输入日期指定时间间隔的日期

`type: function`

```
dateByInterval(interval, start)
```

- `@param/interval`: `type: number` `default: 0` 相隔多少天（负数表示之前的日期）
- `@param/start`: `type: string` 开始日期，默认为今天的日期
- `@return`: `type: string` 指定时间

```
dateByInterval(10, '2019-01-10')                 // '2019-01-20'
dateByInterval(-10, '2019-01-22')                 // '2019-01-12'
```

## relativeToToday：相对今天的日期（当天，过去，未来 & 相隔天数）

`type: function`

```
relativeToToday(date)
```

- `@param/date`: `type: string` 日期
- `@return`: `type: {}` 相对关系对象

```
// today: 2019-01-10
relativeToToday('2019-01-10')               // {type: 0, distance: 0}
relativeToToday('2019-01-12')               // {type: 1, distance: 2}
relativeToToday('2019-01-08')               // {type: -1, distance: 2}
```

## validStartEndDate：检查结束日期是否大于等于开始日期

`type: function`

```
validStartEndDate(start, end)
```

- `@param/start`: `type: string` 开始日期，格式 2019-01-01
- `@param/end`: `type: string` 结束日期，格式 2019-01-01
- `@return`: `type: bool` 是否合法

```
validStartEndDate('2019-01-10')                  // true
validStartEndDate('2019-01-10', '2019-01-10')    // true
validStartEndDate('2019-01-10', '2019-01-12')    // true
validStartEndDate('2019-01-10', '2019-01-08')    // false
```

## htmlToText：提取富文本中的纯文字

`type: function`

```
htmlToText(content)
```

- `@param/content`: `type: string` 富文本
- `@return`: `type: string` 纯文字

```
htmlToText('<p>haha</p>')              // 'haha'
```

## removeATarget：移除 a 标签的 target=\_blank 属性

`type: function`

```
removeATarget(content)
```

- `@param/content`: `type: string` 富文本
- `@return`: `type: string`

```
removeATarget('<a href="/" target=_blank>haha</a>')              // '<a href="/">haha</a>'
```

## readableNumber：格式化为可读数字

`type: function`

```
readableNumber(num, group)
```

- `@param/num`: `type: number` 待处理的数字
- `@param/group`: `type: number` 分组，以多少个数字为一组，默认 3
- `@return`: `type: string`

```
readableNumber(123456789)              // '123,456,789'
readableNumber(123456789.123, 4)              // '1,2345,6789.123'
```

## scrollTo：滚动元素到指定位置

`type: function`

```
scrollTo({el, interval, step, y, onComplete, animate})
```

- `@param/el`: `type: dom` 容器元素，默认 `window`
- `@param/interval`: `type: number` 滚动间隔，默认 20 毫秒
- `@param/step`: `type: number` 每次减小的步长，默认 0.2
- `@param/y`: `type: number` y 坐标位置，默认 0
- `@param/onComplete`: `type: function` 到达目标元素的回调函数
- `@param/animate`: `type: bool` 是否使用动画，默认 true

## safeFloat：安全的 float（保留 N 位小数）

`type: function`

```
safeFloat(f, fixed)
```

- `@param/f`: `type: number` 待处理的数字
- `@param/fixed`: `type: fixed` 小数位数，默认 2
- `@return`: `type: float`

```
readableNumber(0.123456)                 // 0.12
readableNumber(0.123456, 4)              // 0.1234
readableNumber(123456789)                // 123456789
```


CHANGELOG.md
# changelog

## 0.1.2

- 修复 `scrollTo` 的 bug.
- 添加 `safeFloat`.
- 更新 `isMiniProgram` 为 `bool`.

## 0.1.1

- 添加 `parseUrlSearch`
- 添加 `scrollTo`
- `makeUrlSearch` 去掉 `encodeURIComponent`

## 0.1.0

- Init.


src/date.js
import { fillZero } from './num';

// 一天的毫秒数
const secondsOfMinute = 60;
const secondsOfHour = 60 * secondsOfMinute;
const secondsOfDay = 24 * secondsOfHour;
const secondsOfMonth = 30 * secondsOfDay;
const secondsOfYear = 12 * secondsOfMonth;
const millisecondsOfDay = 1000 * secondsOfDay;

/**
 * 获取一个时间对象
 *
 * @returns {{date: string, dateTime: string, hours: number, seconds: number, week: number, d: Date, month: number, year: number, minutes: number, time: string, day: number}}
 */
export const getDate = date => {
  const d = date || new Date();
  const year = d.getFullYear(); // 当前4位年份
  const month = d.getMonth() + 1; // 当前月份 1-12
  const day = d.getDate(); // 当前星期 0-6 (日-六)
  const week = d.getDay(); // 当前几号
  const hours = d.getHours(); // 小时
  const minutes = d.getMinutes(); // 分钟
  const seconds = d.getSeconds(); // 秒

  const monthStr = fillZero(month);
  const dayStr = fillZero(day);
  const hoursStr = fillZero(hours);
  const minutesStr = fillZero(minutes);
  const secondsStr = fillZero(seconds);

  return {
    d,
    year,
    month,
    day,
    week,
    hours,
    minutes,
    seconds,
    date: `${year}-${monthStr}-${dayStr}`,
    time: `${hoursStr}:${minutesStr}:${secondsStr}`,
    dateTime: `${year}-${monthStr}-${dayStr} ${hoursStr}:${minutesStr}:${secondsStr}`,
  };
};

/**
 * 当前时间数据
 */
export const now = getDate();

/**
 * 刷新当前时间数据
 */
export const refreshNow = date => {
  const newDate = getDate(date);

  Object.keys(newDate).forEach(key => {
    now[key] = newDate[key];
  });
};

/**
 * 把日期转成数字
 *
 * '2019-01-01' => 20190101
 *
 * @param date
 * @returns {number}
 */
export const numOfDate = date =>
  parseInt(`${date.slice(0, 4)}${date.slice(5, 7)}${date.slice(8, 10)}`, 10);

/**
 * 把时间转成数字
 *
 * '12:12:12' => 121212
 *
 * @param time
 * @returns {number}
 */
export const numOfTime = time =>
  parseInt(`${time.slice(0, 2)}${time.slice(3, 5)}${time.slice(6, 8)}`, 10);

/**
 * 把日期时间转成数字
 *
 * '2019-01-01 01:01:01' => 20190101010101
 *
 * @param dateTime
 * @returns {number}
 */
export const numOfDateTime = dateTime =>
  parseInt(
    `${dateTime.slice(0, 4)}${dateTime.slice(5, 7)}${dateTime.slice(
      8,
      10,
    )}${dateTime.slice(11, 13)}${dateTime.slice(14, 16)}${dateTime.slice(
      17,
      19,
    )}`,
    10,
  );

/**
 * 返回两个日期之间的距离，以天计
 *
 * @param start
 * @param end
 * @returns {number}
 */
export const distanceOfDates = (start, end) => {
  if (!start || !end) return 0;

  // 兼容 ios
  const startDate = new Date(start.replace(/-/g, '/'));
  const endDate = new Date(end.replace(/-/g, '/'));

  return Math.floor(
    (endDate.getTime() - startDate.getTime()) / millisecondsOfDay,
  );
};

/**
 * 格式化为可读时间
 *
 * 秒前
 * 分钟前
 * 小时前
 * 天前
 * 个月前
 * 年前
 *
 * @param time
 * @returns {string}
 */
export const readableTime = time => {
  if (!time) return '';

  // ios 兼容
  const date = new Date(time.replace(/-/g, '/'));
  const current = new Date();

  const distance = current.getTime() - date.getTime();
  const seconds = Math.floor(distance / 1000);

  if (seconds < secondsOfMinute) return `${seconds}秒前`;
  if (seconds < secondsOfHour)
    return `${Math.floor(seconds / secondsOfMinute)}分钟前`;
  if (seconds < secondsOfDay)
    return `${Math.floor(seconds / secondsOfHour)}小时前`;
  if (seconds < secondsOfMonth)
    return `${Math.floor(seconds / secondsOfDay)}天前`;
  if (seconds < secondsOfYear)
    return `${Math.floor(seconds / secondsOfMonth)}个月前`;
  return `${Math.floor(seconds / secondsOfYear)}年前`;
};

/**
 * 获取距离输入日期指定时间间隔的日期
 *
 * @param interval 相隔多少天（负数表示之前的日期）
 * @param start 开始日期，默认为今天的日期
 * @returns {string}
 */
export const dateByInterval = (interval = 0, start = now.date) => {
  const startDate = new Date(start.replace(/-/g, '/'));
  // 时间戳
  const startDateTime = startDate.getTime();

  const dateTime = startDateTime + interval * millisecondsOfDay;
  const date = new Date(dateTime);
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();

  return `${year}-${fillZero(month)}-${fillZero(day)}`;
};

/**
 * 相对今天的日期（当天，过去，未来 & 相隔天数）
 *
 * @param date
 * @returns {{type: number, distance: number}}
 *     type: -1(past), 0(today), 1(future)
 *     distance: days interval
 */
export const relativeToToday = date => {
  const dates = date.split('-');
  const year = parseInt(dates[0], 10);
  const month = parseInt(dates[1], 10);
  const day = parseInt(dates[2], 10);

  // today
  if (year === now.year && month === now.month && day === now.day)
    return { type: 0, distance: 0 };

  // future
  if (
    year > now.year ||
    (year === now.year && month > now.month) ||
    (year === now.year && month === now.month && day > now.day)
  )
    return { type: 1, distance: distanceOfDates(now.date, date) };

  // past
  return { type: -1, distance: distanceOfDates(date, now.date) };
};

/**
 * 检查结束日期是否大于等于开始日期
 *
 * @param start 开始日期，格式 2019-01-01
 * @param end 结束日期，格式 2019-01-01
 * @returns {boolean}
 */
export const validStartEndDate = (start, end) => {
  if (!start || !end) return !0;

  const startNum = numOfDate(start);
  const endNum = numOfDate(end);

  return endNum >= startNum;
};


src/func.js
/**
 * 重新加载微信浏览器的标题，hack 在微信 webview 中无法修改 document.title 的情况
 *
 * @param title
 */
export const reloadTitle = title => {
  if (!title) return;

  document.title = title;

  const iframe = document.createElement('iframe');
  iframe.style.display = 'none';
  iframe.setAttribute('src', '/favicon.ico');

  iframe.addEventListener('load', () => {
    setTimeout(() => {
      iframe.remove();
    }, 0);
  });

  document.body.appendChild(iframe);
};

/**
 * make url params search
 *
 * @example
 *
 * ```
 * {key1: 1, key2: 2} -> key1=1&key2=2
 * ```
 *
 * @param params
 * @returns {string}
 */
export const makeUrlSearch = params => {
  const urlParams = [];

  Object.keys(params).forEach(key => {
    urlParams.push(`${key}=${params[key] || ''}`);
  });

  return urlParams.join('&');
};

/**
 * parse url search
 *
 * @example
 *
 * ```
 * (?)key1=1&key2=2 -> {key1: 1, key2: 2}
 * ```
 *
 * @param search
 * @returns {string}
 */
export const parseUrlSearch = search => {
  const params = {};

  if (search) {
    const str = search.slice(0, 1) === '?' ? search.slice(1) : search;
    str.split('&').forEach(item => {
      const items = item.split('=');
      params[items[0]] = items[1] || '';
    });
  }

  return params;
};

/**
 * 元素距离文档顶部的位置信息
 *
 * {x, y}
 *
 * @param element
 * @returns {{x: number, y: number}}
 */
export const posToDoc = element => {
  let el = element;

  let x = 0;
  let y = 0;

  while (el) {
    x += el.offsetLeft;
    y += el.offsetTop;
    el = el.offsetParent;
  }

  return { x, y };
};

/**
 * 加载图片快捷函数
 *
 * @param images
 * @param cb
 */
export const loadImages = (images, cb) => {
  const imagesToLoad = Array.isArray(images) ? [...images] : [images];

  let count = 0;
  const load = () => {
    count += 1;
    if (count >= imagesToLoad.length && cb) cb();
  };

  imagesToLoad.forEach(image => {
    const img = new Image();

    img.src = image;
    img.onload = load;
  });
};

// 用于记录上一次位置
const scrollToLastTargetRecord = {
  // id: target
};
let scrollToCount = 1;
/**
 * 滚动元素到指定位置
 *
 * @param el 元素，默认 window
 * @param interval 滚动间隔，默认 20 毫秒
 * @param step 每次减小的步长，默认 0.2
 * @param y y 坐标位置，默认 0
 * @param onComplete 到达目标元素的回调函数
 * @param animate 是否使用动画，默认 true
 */
export const scrollTo = ({
  el = window,
  interval = 20,
  step = 0.2,
  y = 0,
  onComplete,
  animate = !0,
}) => {
  const id = scrollToCount;
  scrollToCount += 1;
  scrollToLastTargetRecord[id] = -1;

  const isWin = el === window;
  const reverseStep = 1 - step;

  const initScrollY = isWin ? el.scrollY : el.scrollTop;
  // 是否是向下滚动
  const isDown = y > initScrollY;

  // 不使用动画
  if (!animate) {
    el.scrollTo(0, y);
    if (onComplete) onComplete();
    return;
  }

  const timer = setInterval(() => {
    const scrollY = isWin ? el.scrollY : el.scrollTop;

    if (!isDown && Math.abs(scrollY - y) < 1) {
      clearInterval(timer);
      if (onComplete) onComplete();
      return;
    }

    const target = !isDown
      ? (scrollY - y) * reverseStep + y
      : Math.ceil((y - scrollY) * step) + scrollY;

    // 如果是向下滚动，元素不够高，无法到达元素位置，则需要此判断，否则会陷入死循环
    if (isDown && scrollToLastTargetRecord[id] === target) {
      clearInterval(timer);
      if (onComplete) onComplete();
      return;
    }

    scrollToLastTargetRecord[id] = target;

    el.scrollTo(0, target);
  }, interval);
};

/**
 * 安全的 float（保留N位小数）
 *
 * @param f
 * @param fixed
 * @returns {number}
 */
export const safeFloat = (f, fixed = 2) => parseFloat(f.toFixed(fixed));


src/num.js
/* eslint-disable import/prefer-default-export */

/**
 * 填充 0
 *
 * @param num
 * @param length 长度，默认为2
 * @returns {string}
 */
export const fillZero = (num = 0, length = 2) => {
  let result = `${num}`;

  while (true) {
    if (result.length >= length) break;

    result = `0${result}`;
  }

  return result;
};

/**
 * 格式化为可读数字
 *
 * 123456789 => 123,456,789
 * 123456789.123 => 1,2345,6789.123
 *
 * @param num 待处理的数字
 * @param group 分组，以多少个数字为一组，默认 3
 * @returns {string}
 */
export const readableNumber = (num = 0, group = 3) => {
  if (!num) return '0';

  const items = `${num}`.split('.');

  const intStr = `${items[0]}`;
  const decimalStr = `${items[1] || ''}`;

  // 不需要进行分组
  if (intStr.length <= group) return `${num}`;

  const reverseIntStr = intStr
    .split('')
    .reverse()
    .join('');
  const resultIntStr = reverseIntStr
    .replace(new RegExp(`(\\d{${group}})`, 'g'), '$1,')
    .split('')
    .reverse()
    .join('');

  const finalIntStr =
    resultIntStr.slice(0, 1) === ',' ? resultIntStr.slice(1) : resultIntStr;

  return `${finalIntStr}${decimalStr ? `.${decimalStr}` : ''}`;
};
